pg_resgroup_move_query() improvements.

InnerLife0 opened the pull request:
pg_resgroup_move_query() improvements.
The pg_resgroup_move_query() function implemented in 51ee26b has several disadvantages:
1) Slot leaking. The slot acquired in initiator process doesn't free if target process isn't alive, didn't receive a signal, or even received, but at the time it was in idle state.
2) Race condition between UnassignResGroup() called at the end of transaction (or honestly any other code) and handler called immediately after USR1 signal.
3) Not working entrydb process moving. Previously, the USR1 signal was sent to only one process, target or entrydb. Entrydb moving was never tested properly.

Improvements added to solve the first problem:
1) Feedback from target process to initiator. Target process can set initiator's latch to quickly interrupt pg_resgroup_move_query() from awaiting.
2) Existed movetoResSlot parameter acts like a mark. Initiator sets it and waits on latch. If movetoResSlot become NULL, slot control is on the target process side.
3) Initiator PID. Used by target process to get initiator's latch.
4) Mutex. To guard all critical moveto* parameters from parallel changes.

To solve the second problem, there was an attempt to use Postgres interruptions. I was unhappy to know GPDB use raw InterruptPending value to do some pre-cancellation of dispatched queries. GPDB thinks InterruptPending can be triggered only by "negative" events, which leads to cancelation. The temporary solution with InterruptPending-like "positive" flag showed that we may wait for next CHECK_FOR_INTERRUPTS() call for a long. For example, some fat insert may do group moving only at the end of dispatching, which makes no sense.
Thus, I decided to decrease the probability of races by using additional IsTransactionState() check.
IMPORTANT NOTICE. The existed solution still affected by race conditions. The current handler's implementation is an example of bad design and should be reworked.

To solve the third problem, now we send signals to target and entrydb processes separately. We do UnassignResGroup() for one process only, not cleaning all counters for still working entrydb process inside handler. More, entrydb moving logic is now separated from segments moving, and so, it became more readable.

New GUC (gp_resource_group_move_timeout) was added to limit a time we're waiting for feedback from target.

New regression tests shows there is no slot leaking with some rare and now correctly resolved cases.

InnerLife0 commented on src/backend/utils/resgroup/resgroup.c:
```diff
@@ -4766,6 +4796,78 @@ moveQueryCheck(int sessionId, Oid groupId)
+		 * no much difference of why process wasn't moved (by timeout or by
+		 * death reasons), but the logic around became more complex.
+		 */
+		WaitLatch(&MyProc->procLatch,
```
Curious to know do we have some pitfalls not analyzing `WL_POSTMASTER_DEATH` return status like it's done [here](https://github.com/arenadata/gpdb/blob/c1a3ea3c2af5d329fdb012abc18c8ca42f5612b7/src/backend/utils/resgroup/resgroup.c#L2940).

Stolb27 commented on src/backend/utils/resgroup/resgroup.c:
```diff
@@ -4766,6 +4809,49 @@ moveQueryCheck(int sessionId, Oid groupId)
+
+	for (;;)
+	{
+		ResetLatch(&MyProc->procLatch);
```
What if the target process has already set latch?

Stolb27 commented on src/backend/utils/resgroup/resgroup.c:
```diff
@@ -4766,6 +4809,49 @@ moveQueryCheck(int sessionId, Oid groupId)
+			break;
+		if (latchRes & WL_POSTMASTER_DEATH)
+			elog(ERROR, "got WL_POSTMASTER_DEATH waiting on latch; exiting...");
+	}
```
Do we go to the next iteration in the case of WL_LATCH_SET?

We need to check our condition (remote process handle signal) before leaving this loop. This latch may be set in many other places for different purposes (e.g. in SIGHUP [handler](https://github.com/arenadata/gpdb/blob/d4534e08c27ede9b572ecfcf0f167518bf562ff4/src/backend/tcop/postgres.c#L3614) to reread config). And, as result, we may leave this loop before time is running up and the target process handles the signal.

Stolb27 commented on src/backend/storage/ipc/procarray.c:
```diff
@@ -4943,24 +4949,116 @@ ResGroupSignalMoveQuery(int sessionId, void *slot, Oid groupId)
+		 * session
+		 */
+		if (Gp_role == GP_ROLE_DISPATCH)
+			break;
```
What if it was entrydb process on coordinator with the same session id?

Stolb27 commented on src/backend/utils/resgroup/resgroup.c:
```diff
@@ -4617,38 +4622,75 @@ HandleMoveResourceGroup(void)
+		SpinLockAcquire(&MyProc->movetoMutex);
+		/* notify initiator if command still relevant */
+		if (MyProc->movetoResSlot)
+			(void) ResGroupMoveNotifyInitiator(MyProc->movetoCallerPid);
```
Is this operation short enough to hold spinlock?

InnerLife0 replied:
With this simplified version, we need to do `ResGroupMoveNotifyInitiator()` and `MyProc->movetoResSlot = NULL` inside one lock or initiator may read incorrect MyProc->movetoResSlot result.

Stolb27 commented on src/backend/utils/resgroup/resgroup.c:
```diff
@@ -4766,6 +4816,81 @@ moveQueryCheck(int sessionId, Oid groupId)
+					elog(ERROR,
+						"got WL_POSTMASTER_DEATH waiting on latch; exiting...");
+			}
+			PG_CATCH();
```
Which kind of errors do you expect here and in the test case? Do you expect WaitLatch ERRORs? Am I right?
The logic is ok, but wrapping WaitLatch in try/catch isn't distributed spreadly in gpdb code base.

InnerLife0 replied:
> Which kind of errors do you expect here and in the test case? Do you expect WaitLatch ERRORs? Am I right? 

Yes.

> The logic is ok, but wrapping WaitLatch in try/catch isn't distributed spreadly in gpdb code base.

Probably, because there nothing shared to free in another cases.
Here we need to clean all flags and get the actual result of handling to decide can we free slot or not.

Stolb27 commented on src/backend/storage/ipc/procarray.c:
```diff
@@ -4943,24 +4949,156 @@ ResGroupSignalMoveQuery(int sessionId, void *slot, Oid groupId)
+			}
 		}
+		SpinLockRelease(&proc->movetoMutex);
+		break;
```
What about entrydb here? You set fields for both of possible processes, but validate only the first of them.
entrydb reset only groupid as executor.

InnerLife0 replied:
And here we back to the question of how to distinguish entrydb.
One of the possible solutions for now - run over all procArray and keep "the most successful" results.

Stolb27 commented on src/test/isolation2/input/resgroup/resgroup_move_query.source:
```diff
@@ -141,9 +151,134 @@ SELECT is_session_in_group(pid, 'rg_move_query_mem_small') FROM pg_stat_activity
+2: SELECT gp_inject_fault('resource_group_give_away3', 'reset', dbid) FROM gp_segment_configuration where role = 'p' and content = -1;
+2: SELECT gp_inject_fault('resource_group_give_away3', 'suspend', dbid) FROM gp_segment_configuration where role = 'p' and content = -1;
+2&: SELECT gp_toolkit.pg_resgroup_move_query(pid, 'rg_move_query') FROM pg_stat_activity WHERE query LIKE '%pg_sleep%' AND rsgname='rg_move_query_mem_small';
+3: SELECT pg_sleep(4);
```
Why do you expect that the 2nd session already achieve fault injector? Do we decrease default timeout from 30 seconds?

InnerLife0 replied:
I expect you are right.

hughcapet commented on gpcontrib/gp_internal_tools/gp_resource_group.c:
```diff
@@ -84,6 +84,11 @@ pg_resgroup_move_query(PG_FUNCTION_ARGS)
 
+		if (pid == MyProcPid)
+			ereport(ERROR,
+					(errcode(ERRCODE_UNDEFINED_OBJECT),
```
Isn't `ERRCODE_INVALID_PARAMETER_VALUE` better here?

InnerLife0 replied:
Yes, it is. Thanks.

InnerLife0 replied:
Fixed.

hughcapet commented on src/backend/storage/ipc/procarray.c:
```diff
@@ -4929,38 +4929,188 @@ GetSessionIdByPid(int pid)
+ * Cleaning and checking should be performed as one atomic operation inside one
+ * mutex.
+ * 'clean' flag is bidirectional. If 'clean' is set to true, then all moveto*
+ * params will be cleaned, no matter was target handled them or not.
+ * At the other side, it will be forcefully set to true, if target process
```
```suggestion
 * params will be cleaned, no matter has target already handled them or not.
 * On the other side, it will be forcefully set to true, if target process
```

InnerLife0 replied:
Can't understand why we need to change first line.
As for second, "on the other side"/"at the other hand" mixed up in my mind. But they both don't fit well here. I'll fix this sentence. Thanks.

InnerLife0 replied:
Changed second line.
First line left unchanged.

hughcapet commented on src/backend/utils/resgroup/resgroup.c:
```diff
@@ -4766,6 +4828,81 @@ moveQueryCheck(int sessionId, Oid groupId)
+	}
+
+	if (!res)
+		elog(ERROR, "target process can't move itself to a new group");
```
I think 'failed to move' is more clear, isn't it?

InnerLife0 replied:
I don't see any big difference, but it's shorter at least. OK.

hughcapet replied:
I should explain: because of "can't" I thought you mean it is an  _invalid_, not failed operation, as sentence is quite similar to https://github.com/arenadata/gpdb/blob/665710722804bda1ad7248a8214c70f6ee29dd06/gpcontrib/gp_internal_tools/gp_resource_group.c#L87-L90

~dushnila~

hughcapet replied:
And I guess it is better to use `ereport` in the new code.

InnerLife0 replied:
Still can't understand the play around "can't".  Quote: "Can’t is a contraction of cannot...", so they are the same.
There is a thing which may be interesting and makes sense: "...and as such it’s sometimes unsuitable for formal writing.", but it seems that's not what you meant.

InnerLife0 replied:
Moved from "can't" to "cannot" to be more formal.
`elog` replaced by "ereport" where it was possible.

hughcapet commented on src/backend/storage/ipc/procarray.c:
```diff
@@ -4929,38 +4929,188 @@ GetSessionIdByPid(int pid)
+		 * As a mediocre solution we assume executors (including entrydb) not
+		 * using queryCommandId.
+		 */
+		if ((proc->queryCommandId <= -1) ^ executor)
```
Although I don't have any alternatives to suggest, don’t like the idea of relying on the queryCommandId identifier. There is not a zero percent chance someday we will need it on executors as well

InnerLife0 replied:
The guard is [here](https://github.com/arenadata/gpdb/pull/300/files/665710722804bda1ad7248a8214c70f6ee29dd06#diff-14f7111e45936f7f6b0fbccf732d91ab7957e461b8a3e641381e1b003f113a18R4711-R4716).

hughcapet replied:
Agree, although still don't like it :)

InnerLife0 replied:
The word "mediocre" points to this too.

hughcapet replied:
Hm... And what do you think about mppIsWriter property?
It is seems, it is [always true](https://github.com/arenadata/gpdb/blob/ADBDEV-2419-2/src/backend/storage/lmgr/proc.c#L490-L493) for dispatcher role
And is always false for [entrydb](https://github.com/arenadata/gpdb/blob/ADBDEV-2419-2/src/backend/cdb/cdbutil.c#L805) (for [utility role as well](https://github.com/arenadata/gpdb/blob/ADBDEV-2419-2/src/backend/cdb/cdbvars.c#L471-L472))

InnerLife0 replied:
OK. So it should be less tricky, right?

hughcapet replied:
I suppose, it is will look almost the same but slightly less mediocre (there are some places in code where the same approach is used), what is your opinion?

InnerLife0 replied:
Not so handy as `queryCommandId`, but OK.
I'll try to use it.

InnerLife0 replied:
Switched to use of `mppIsWriter`. I proved the behavior by debugging.
Assert() in handler removed.
Thanks.

hughcapet commented on src/backend/storage/ipc/procarray.c:
```diff
@@ -4929,38 +4929,188 @@ GetSessionIdByPid(int pid)
-		if (Gp_role == GP_ROLE_DISPATCH)
+
+		SpinLockAcquire(&proc->movetoMutex);
+		if (Gp_role == GP_ROLE_DISPATCH && proc->queryCommandId > -1)
```
Maybe `if (!executor)` can also work here?

InnerLife0 replied:
Here and in functions near or in another files we rely on `proc->queryCommandId`, but not on a local `executor` argument, so it's cleaner to keep it's usage if possible.

hughcapet replied:
Well... This function seemed to me to be overcomplicated: passed func argument 'executor' + local definition of 'executor' +  proc->queryCommandId comparison
But it is probably the only way to do it safely, ~who am I~ I think colleagues know better

InnerLife0 replied:
Yes, it's overcomplicated. But I don't see any good option to move some logic to another (new) one.
UPD: I'll add additional comment to function description.

InnerLife0 replied:
Additional comment added. The logic still the same.

hughcapet commented on src/backend/utils/resgroup/resgroup.c:
```diff
@@ -4801,19 +4937,48 @@ ResGroupMoveQuery(int sessionId, Oid groupId, const char *groupName)
-		PG_RE_THROW();
+		/* 
+		 * we don't have proper mechanics to cancel group move,
+		 * so just warn about something wrong on dispaching stage
```
```suggestion
		 * so just warn about something went wrong on dispatching stage
```

InnerLife0 replied:
Thanks for typo.
As for "went". I think there is no difference in this context.

hughcapet replied:
Yes, agree. I have misread it first, sorry

hughcapet commented:
Found something nasty while emulating pg_resgroup_move_query caller cancellation:
```
SELECT gp_inject_fault('resource_group_give_away1', 'interrupt', dbid) FROM gp_segment_configuration where role = 'p' and content = -1;
1: begin;
2: select gp_toolkit.pg_resgroup_move_query(<pid_1>, <necessary_group>); 
ERROR:  canceling statement due to user request
1: select 1;
FATAL:  Unexpected internal error (resgroup.c:3280)
DETAIL:  FailedAssertion("!(groupIsNotDropped(group))", File: "resgroup.c", Line: 3280)
```

```
#4  0x0000000000c8d00d in ExceptionalCondition (conditionName=conditionName@entry=0x13e4942 "!(groupIsNotDropped(group))", errorType=errorType@entry=0x11e01e4 "FailedAssertion", 
    fileName=fileName@entry=0x13e4725 "resgroup.c", lineNumber=lineNumber@entry=3280) at assert.c:46
#5  0x0000000000ce2349 in selfSetGroup (group=0x0) at resgroup.c:3280
#6  selfAttachResGroup (group=0x0, slot=slot@entry=0x7f1420dfe890) at resgroup.c:1548
#7  0x0000000000ce9dbe in HandleMoveResourceGroup () at resgroup.c:4693
```

InnerLife0 commented:
> Found something nasty while emulating pg_resgroup_move_query caller cancellation:

One more found bug in resgroup moving, Yay!
I'll take a look and check should we add a fix of new bug to current patch.

maksm90 commented on gpdb-doc/dita/admin_guide/workload_mgmt_resgroups.xml:
```diff
@@ -1084,7 +1084,14 @@ gpstart
+        <li>If the group has a free slot, <codeph>pg_resgroup_move_query()</codeph> tries
+          to give slot control away to the target process for up to
+          <codeph>gp_resource_group_move_timeout</codeph> milliseconds. If target process
+          can't handle move signal until <codeph>gp_resource_group_queuing_timeout</codeph>
```
*gp_resource_group_move_timeout

> can't handle move signal

from user's perspective it's not evident that we actually send signal under the hood then I propose to rename "move signal" to "move request".

InnerLife0 replied:
Thanks

maksm90 commented on gpdb-doc/dita/ref_guide/config_params/guc-list.xml:
```diff
@@ -4724,6 +4725,38 @@
+      <p>Cancel movement of process to another resource group if target process
+        not replies longer than the specified number of milliseconds. This parameter
+        doesn't include <codeph>gp_resource_group_queuing_timeout</codeph> - queuing
+        on resource group performed before sending movement signal.</p>
```
similar note to "movement signal"

InnerLife0 replied:
Thanks

maksm90 commented on src/test/isolation2/input/resgroup/resgroup_move_query.source:
```diff
@@ -91,6 +93,7 @@ CREATE ROLE role_move_query_mem_small RESOURCE GROUP rg_move_query_mem_small;
 1: SELECT hold_memory_by_percent(1,1.0);
 SELECT gp_toolkit.pg_resgroup_move_query(pid, 'rg_move_query_mem_small') FROM pg_stat_activity WHERE query LIKE '%hold_memory_by_percent%' AND state = 'idle in transaction';
 SELECT is_session_in_group(pid, 'rg_move_query_mem_small') FROM pg_stat_activity WHERE query LIKE '%hold_memory_by_percent%' AND state = 'idle in transaction';
+SELECT num_running FROM gp_toolkit.gp_resgroup_status WHERE rsgname='rg_move_query_mem_small';
```
Why have you added this query semantically similar to the previous one?

_Off-top_. Taking into account that session movement is asynchronous, after issuing of `gp_toolkit.pg_resgroup_move_query()` we have to wait some time before starting of verification that processes are in target resource group. Ideally, we have to wait until some injector point will be achieved.

InnerLife0 replied:
`is_session_in_group()` and `num_running` checks different things. `is_session_in_group()` checks (literally) all the processes moved to new group. Line with `num_running` checks we have no slot leaking after unsuccessful ('f' as result of previous line) attempt of moving.
As this test is negative and we know it proves working of some checks logically placed before sending signal, we can ignore awaiting (we have nothing to await).

maksm90 commented on src/backend/utils/resgroup/resgroup.c:
```diff
@@ -4766,28 +4823,106 @@ moveQueryCheck(int sessionId, Oid groupId)
+	bool		clean = false;
+	bool		res = false;
+
+	SIMPLE_FAULT_INJECTOR("resource_group_give_away1");
```
This fault point might be moved on one level up and renamed to smth like "resource_group_after_acquiring_before_give_away" to clarify its position in corresponding tests.

InnerLife0 replied:
Or it can be renamed to `resource_group_give_away_begin` then, without moving. It doesn't matter where it placed in such case.

InnerLife0 replied:
Fault injection points were renamed.

maksm90 commented on src/test/isolation2/input/resgroup/resgroup_move_query.source:
```diff
@@ -141,9 +145,150 @@ SELECT is_session_in_group(pid, 'rg_move_query_mem_small') FROM pg_stat_activity
+-- wait until first process finish
+-- resume at second process, it should throw an error
+1: SET ROLE role_move_query_mem_small;
+1&: SELECT pg_sleep(3);
```
In general, I don't endorse the concurrency control via `pg_sleep()` call despite of in previous cases (that are implemented far non-perfectly) this method is used. I propose to find some universal fault point inside transaction block and use it in the flow of first process.

InnerLife0 replied:
`pg_sleep()` acts here like a 'long-running query', not to control concurrency.
If I understand you right, you want to find or add an additional point somewhere in the end of transaction. Some point, where transaction already treated as idle. In this case first process can execute any query (not necessary long-running) and stop on this point awaiting for second process stuff.

InnerLife0 replied:
pg_sleep() left only for 'long-running query' simulation.

maksm90 commented on src/test/isolation2/input/resgroup/resgroup_move_query.source:
```diff
@@ -141,9 +145,150 @@ SELECT is_session_in_group(pid, 'rg_move_query_mem_small') FROM pg_stat_activity
+2: SELECT gp_inject_fault('resource_group_give_away2-2', 'reset', dbid) FROM gp_segment_configuration where role = 'p' and content = -1;
+2: SELECT gp_inject_fault('resource_group_give_away2-2', 'error', dbid) FROM gp_segment_configuration where role = 'p' and content = -1;
+2&: SELECT gp_toolkit.pg_resgroup_move_query(pid, 'rg_move_query') FROM pg_stat_activity WHERE query LIKE '%pg_sleep%' AND rsgname='rg_move_query_mem_small';
+3: SELECT pg_sleep(3);
```
Instead of `pg_sleep()` call we might use multiple `gp_wait_until_triggered_fault()` ones to wait for passing of all interesting fault points.

InnerLife0 replied:
Makes sense. Thanks.

InnerLife0 replied:
Fixed in all tests where it was possible.

hughcapet commented on src/backend/storage/ipc/procarray.c:
```diff
@@ -4928,39 +4928,193 @@ GetSessionIdByPid(int pid)
+ * More, it will be forcefully set to true, if target process handled our
+ * command. Thus, if function returned true in 'clean', it should be treated
+ * as terminal state and all new calls to ResGroupMoveCheckTargetReady()
+ * before calling ResGroupMoveSignalTarget() makes no sense.
```
```suggestion
 * before calling ResGroupMoveSignalTarget() make no sense.
```

InnerLife0 replied:
Thanks

hughcapet commented on src/backend/utils/resgroup/resgroup.c:
```diff
@@ -3386,7 +3395,12 @@ slotValidate(const ResGroupSlotData *slot)
 		Assert(!slotIsInFreelist(slot));
-		AssertImply(Gp_role == GP_ROLE_EXECUTE, slot == sessionGetSlot());
+		/*
+		 * Entrydb process can has different self and session slots at the
```
```suggestion
		 * Entrydb process can have different self and session slots at the
```

InnerLife0 replied:
Thanks

hughcapet commented on src/backend/utils/resgroup/resgroup.c:
```diff
@@ -4766,28 +4876,110 @@ moveQueryCheck(int sessionId, Oid groupId)
+					*slot = NULL;
+					ereport(WARNING,
+							(errmsg("got exception, but slot control is on the target process side"),
+							 errhint("QEs not moved. They'll move at next command dispatched in target's transaction, if any.")));
```
```suggestion
							 errhint("QEs weren't moved. They'll be moved by the next command dispatched in the target transaction, if any.")));
```

InnerLife0 replied:
Thanks

hughcapet commented:
It probably should not be a part of this PR but don't we want to return false/raise an error from `pg_resgroup_move_query` if it is called in `gp_session_role=utility`? It now looks like we have sent the signal successfully, although it is not the actual state of things.

hughcapet reviewed (commented):
Sorry, dushn'ila is back in the game

InnerLife0 commented:
> It probably should not be a part of this PR but don't we want to return false/raise an error from `pg_resgroup_move_query` if it is called in `gp_session_role=utility`? It now looks like we have sent the signal successfully, although it is not the actual state of things. And maybe we need some tests for gp_role=utility...? (both utility as a target and a source process)

pg_resgroup_move_query() has nothing to do when role is not GP_ROLE_DISPATCH and not GP_ROLE_EXECUTE.
As decided in voice chat, we left this function return true in utility mode.

Stolb27 commented on src/backend/storage/ipc/procarray.c:
```diff
@@ -4928,39 +4928,193 @@ GetSessionIdByPid(int pid)
  * slot is NULL on QE.
+ * The process with Gp_role == GP_ROLE_DISPATCH can act as executor in case of
+ * entrydb. 'isExecutor' helps us to determine a process to which we need to
+ * send signal.
```
@InnerLife0, this comment is a bit confusing. entry db process has role GP_ROLE_EXECUTE. As far I understand, you said about the possibility to distinguish the target process role with the given flag.

InnerLife0 replied:
That's true, it's a mistake. Fixed.

Commit:
pg_resgroup_move_query() improvements.
The pg_resgroup_move_query() function implemented in 51ee26b has several disadvantages:
1) Slot leaking. The slot acquired in initiator process doesn't free if target process isn't alive, didn't receive a signal, or even received, but at the time it was in idle state.
2) Race condition between UnassignResGroup() called at the end of transaction (or honestly any other code) and handler called immediately after USR1 signal.
3) Not working entrydb process moving. Previously, the USR1 signal was sent to only one process, target or entrydb. Entrydb moving was never tested properly.

Improvements added to solve the first problem:
1) Feedback from target process to initiator. Target process can set initiator's latch to quickly interrupt pg_resgroup_move_query() from awaiting.
2) Existed movetoResSlot parameter acts like a mark. Initiator sets it and waits on latch. If movetoResSlot become NULL, slot control is on the target process side.
3) Initiator PID. Used by target process to get initiator's latch.
4) Mutex. To guard all critical moveto* parameters from parallel changes.

To solve the second problem, there was an attempt to use Postgres interruptions. I was unhappy to know GPDB use raw InterruptPending value to do some pre-cancellation of dispatched queries. GPDB thinks InterruptPending can be triggered only by "negative" events, which leads to cancelation. The temporary solution with InterruptPending-like "positive" flag showed that we may wait for next CHECK_FOR_INTERRUPTS() call for a long. For example, some fat insert may do group moving only at the end of dispatching, which makes no sense.
Thus, I decided to decrease the probability of races by using additional IsTransactionState() check.
IMPORTANT NOTICE. The existed solution still affected by race conditions. The current handler's implementation is an example of bad design and should be reworked.

To solve the third problem, now we send signals to target and entrydb processes separately. We do UnassignResGroup() for one process only, not cleaning all counters for still working entrydb process inside handler. More, entrydb moving logic is now separated from segments moving, and so, it became more readable.

New GUC (gp_resource_group_move_timeout) was added to limit a time we're waiting for feedback from target.

New regression tests shows there is no slot leaking with some rare and now correctly resolved cases.

Commit:
Addressed comments from upstream.
New code comments, asserts, formatting.

Stolb27 committed:
Merge branch 'adb-6.x' into ADBDEV-2419-2

Stolb27 committed:
Merge branch 'adb-6.x-dev' into ADBDEV-2419-2
